services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/security/HankeControllerSecurityTests.kt

```diff
@@ -0,0 +1,179 @@
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+import java.time.ZonedDateTime
+import java.time.temporal.ChronoUnit
+
+
+/**
+ * Tests to ensure HankeController endpoints have correct authentication and
+ * authorization restrictions and handling.
+ *
+ * The actual activities in each test are not important, as long as they cause
+ * the relevant method to be called.
+ * "Perform" function implementations have been adapted from HankeControllerITests.
+ * As long as the mocked operation goes through and returns something "ok" (when
+ * authenticated correctly), it will be usable as an authentication test's operation.
+ *
+ * TODO
+ * Note, HankeService's method-based security is to be tested separately, (in own test class),
+ * once it gets any activated. For now, testing the Controller is enough as the service's
+ * finer things are not implemented.
+ */
+@WebMvcTest(HankeController::class)
+@Import(IntegrationTestConfiguration::class, IntegrationTestResourceServerConfig::class)
+@ActiveProfiles("itest")
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class HankeControllerSecurityTests(@Autowired val mockMvc: MockMvc) {
+
+    @Autowired
+    private lateinit var hankeService: HankeService
+
+    private val testHankeTunnus = "HAI21-TEST-1"
+
+
+    @Test
+    @WithMockUser(username = "test7358", roles = ["haitaton-user"])
+    fun `status ok with authenticated user with correct role`() {
+        performGetHankkeet().andExpect(status().isOk)
+        performPostHankkeet().andExpect(status().isOk)
+        performPutHankkeetTunnus().andExpect(status().isOk)
+        performGetHankkeetTunnus().andExpect(status().isOk)
+    }
+
+    @Test
+    // Without mock user, i.e. anonymous
+    fun `status unauthorized (401) without authenticated user`() {
+        performGetHankkeet()
+                .andExpect(unauthenticated())
+                .andExpect(status().isUnauthorized)
+        performPostHankkeet()
+                .andExpect(unauthenticated())
+                .andExpect(status().isUnauthorized)
+        performPutHankkeetTunnus()
+                .andExpect(unauthenticated())
+                .andExpect(status().isUnauthorized)
+        performGetHankkeetTunnus()
+                .andExpect(unauthenticated())
+                .andExpect(status().isUnauthorized)
+    }
+
+    @Test
+    @WithMockUser(username = "test7358", roles = ["bad role"])
+    fun `status forbidden (403) with authenticated user with bad role`() {
+        performGetHankkeet().andExpect(status().isForbidden)
+        performPostHankkeet().andExpect(status().isForbidden)
+        performPutHankkeetTunnus().andExpect(status().isForbidden)
+        performGetHankkeetTunnus().andExpect(status().isForbidden)
+    }
+
```

Comment:
Is it really testing that user is authenticated like it says in test name? Status.isOk-check is missing so I'm just wondering.. Maybe isForbidden() does that too?

Reply:
Authenticated means the user is recognized (but does not necessarily have the required permissions to what he is requesting). Having bad role or missing role leads to unauthorized situation, which is shown as the forbidden (403) -response.
The status can be (among other things) one of ok (200), forbidden (403) or unauthorized (401, slightly incorrect term used by HTTP). That is, if the status matches isForbidden, it can not match isOk.